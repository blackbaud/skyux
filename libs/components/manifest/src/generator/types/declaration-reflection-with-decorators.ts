import { type DeclarationReflection } from 'typedoc';

/**
 * TypeDoc's `DeclarationReflection` with the addition of decorators.
 * (See 'plugins/typedoc-plugin-decorators.mjs').
 */
export interface DeclarationReflectionWithDecorators
  extends DeclarationReflection {
  decorators?: { name: string; arguments?: Record<string, string> }[];
}
