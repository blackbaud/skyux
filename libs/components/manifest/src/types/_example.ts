import {
  SkyManifestClassDefinition,
  SkyManifestClassMethodDefinition,
  SkyManifestClassPropertyDefinition,
} from './class-def';
import {
  SkyManifestDirectiveDefinition,
  SkyManifestDirectiveInputDefinition,
  SkyManifestDirectiveOutputDefinition,
} from './directive-def';
import {
  SkyManifestEnumerationDefinition,
  SkyManifestEnumerationMemberDefinition,
} from './enumeration-def';
import {
  SkyManifestFunctionDefinition,
  SkyManifestParameterDefinition,
} from './function-def';
import { SkyManifestInterfaceDefinition } from './interface-def';
import { SkyManifestPublicApi } from './manifest';
import { SkyManifestPipeDefinition } from './pipe-def';
import { SkyManifestTypeAliasDefinition } from './type-alias-def';
import { SkyManifestVariableDefinition } from './variable-def';

const classProperty: SkyManifestClassPropertyDefinition = {
  codeExample: '',
  codeExampleLanguage: 'typescript',
  defaultValue: 'baz',
  deprecationReason: '',
  description: 'This is a description.',
  isDeprecated: false,
  isPreview: false,
  kind: 'class-property',
  name: 'bar',
  type: 'string',
};

const classMethod: SkyManifestClassMethodDefinition = {
  codeExample: '',
  codeExampleLanguage: 'typescript',
  deprecationReason: '',
  description: 'This is a description.',
  isDeprecated: false,
  isPreview: false,
  isStatic: false,
  kind: 'class-method',
  name: 'doSomething',
  parameters: [
    {
      defaultValue: '',
      description: '',
      isOptional: false,
      name: 'foo',
      type: 'string',
    },
  ],
  type: 'void',
};

const fooClass: SkyManifestClassDefinition = {
  anchorId: 'sky-foo',
  children: [classProperty, classMethod],
  codeExample: '',
  codeExampleLanguage: 'typescript',
  deprecationReason: '',
  description: 'This is a description.',
  filePath: 'libs/foobar/src/lib/foo.ts',
  isDeprecated: false,
  isInternal: false,
  isPreview: false,
  kind: 'class',
  name: 'SkyFoo',
};

const directiveInput: SkyManifestDirectiveInputDefinition = {
  codeExample: '',
  codeExampleLanguage: 'typescript',
  defaultValue: 'baz',
  deprecationReason: '',
  description: 'This is a description.',
  isDeprecated: false,
  isPreview: false,
  isRequired: false,
  kind: 'directive-input',
  name: 'bar',
  type: 'string',
};

const directiveOutput: SkyManifestDirectiveOutputDefinition = {
  codeExample: '',
  codeExampleLanguage: 'typescript',
  deprecationReason: '',
  description: 'This is a description.',
  isDeprecated: false,
  isPreview: false,
  kind: 'directive-output',
  name: 'barChange',
  type: 'EventEmitter<string>',
};

const directive: SkyManifestDirectiveDefinition = {
  anchorId: 'sky-foo',
  children: [directiveInput, directiveOutput],
  codeExample: '',
  codeExampleLanguage: 'typescript',
  deprecationReason: '',
  description: 'This is a description.',
  filePath: 'libs/foobar/src/lib/foo.ts',
  isDeprecated: false,
  isInternal: false,
  isPreview: false,
  kind: 'directive',
  name: 'SkyFoo',
  selector: '[skyFoo]',
};

const enumMember: SkyManifestEnumerationMemberDefinition = {
  codeExample: '',
  codeExampleLanguage: 'typescript',
  deprecationReason: '',
  description: 'This is a description.',
  isDeprecated: false,
  isPreview: false,
  name: 'bar',
  kind: 'enum-member',
  type: '0',
};

const enumeration: SkyManifestEnumerationDefinition = {
  anchorId: 'sky-foo',
  codeExample: '',
  codeExampleLanguage: 'typescript',
  deprecationReason: '',
  description: 'This is a description.',
  filePath: 'libs/foobar/src/lib/foo.ts',
  isDeprecated: false,
  isInternal: false,
  isPreview: false,
  kind: 'enumeration',
  name: 'SkyFoo',
  children: [enumMember],
};

const fun: SkyManifestFunctionDefinition = {
  anchorId: 'sky-foo',
  codeExample: '',
  codeExampleLanguage: 'typescript',
  deprecationReason: '',
  description: 'This is a description.',
  filePath: 'libs/foobar/src/lib/foo.ts',
  isDeprecated: false,
  isInternal: false,
  isPreview: false,
  kind: 'function',
  name: 'skyFoo',
  parameters: [
    {
      defaultValue: '',
      description: '',
      isOptional: false,
      name: 'foo',
      type: 'string',
    } satisfies SkyManifestParameterDefinition,
  ],
  type: 'void',
};

const int: SkyManifestInterfaceDefinition = {
  anchorId: 'sky-foo',
  codeExample: '',
  codeExampleLanguage: 'typescript',
  deprecationReason: '',
  description: 'This is a description.',
  filePath: 'libs/foobar/src/lib/foo.ts',
  isDeprecated: false,
  isInternal: false,
  isPreview: false,
  kind: 'interface',
  name: 'SkyFoo',
  children: [
    {
      codeExample: '',
      codeExampleLanguage: 'typescript',
      deprecationReason: '',
      description: 'This is a description.',
      isDeprecated: false,
      isPreview: false,
      isOptional: false,
      kind: 'interface-property',
      name: 'bar',
      type: 'string',
    },
  ],
};

const pipe: SkyManifestPipeDefinition = {
  anchorId: 'sky-foo',
  codeExample: '',
  codeExampleLanguage: 'typescript',
  deprecationReason: '',
  description: 'This is a description.',
  filePath: 'libs/foobar/src/lib/foo.ts',
  isDeprecated: false,
  isInternal: false,
  isPreview: false,
  kind: 'pipe',
  name: 'SkyFoo',
  templateBindingName: 'foo',
  children: [classMethod],
};

const typeAlias: SkyManifestTypeAliasDefinition = {
  anchorId: 'sky-foo',
  codeExample: '',
  codeExampleLanguage: 'typescript',
  deprecationReason: '',
  description: 'This is a description.',
  filePath: 'libs/foobar/src/lib/foo.ts',
  isDeprecated: false,
  isInternal: false,
  isPreview: false,
  kind: 'type-alias',
  name: 'SkyFoo',
  type: 'string',
};

const variable: SkyManifestVariableDefinition = {
  anchorId: 'sky-foo',
  codeExample: '',
  codeExampleLanguage: 'typescript',
  deprecationReason: '',
  description: 'This is a description.',
  filePath: 'libs/foobar/src/lib/foo.ts',
  isDeprecated: false,
  isInternal: false,
  isPreview: false,
  kind: 'variable',
  name: 'skyFoo',
  type: 'string',
};

export const publicApi: SkyManifestPublicApi = {
  packages: {
    '@skyux/foobar': [
      fooClass,
      directive,
      enumeration,
      fun,
      int,
      pipe,
      typeAlias,
      variable,
    ],
  },
};
