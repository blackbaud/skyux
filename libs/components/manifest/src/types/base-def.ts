/**
 * The kind of entity exported from the public API.
 * @internal
 */
export type SkyManifestParentDefinitionKind =
  | 'class'
  | 'component'
  | 'directive'
  | 'enumeration'
  | 'function'
  | 'interface'
  | 'module'
  | 'pipe'
  | 'service'
  | 'type-alias'
  | 'variable';

/**
 * The kind of child entity exported from the public API.
 * @internal
 */
export type SkyManifestChildDefinitionKind =
  | 'class-method'
  | 'class-property'
  | 'directive-input'
  | 'directive-output'
  | 'enum-member'
  | 'interface-property';

export type SkyManifestCodeExampleLanguage = 'markup' | 'typescript';

/**
 * Information captured from JSDoc comments.
 * @internal
 */
export interface SkyManifestJsDocDefinition {
  codeExample?: string;
  codeExampleLanguage?: SkyManifestCodeExampleLanguage;
  deprecationReason?: string;
  description?: string;
  isDeprecated?: boolean;
  isPreview?: boolean;
}

/**
 * Information shared by all top-level entities exported from the public API.
 * @internal
 */
export interface SkyManifestParentDefinition
  extends SkyManifestJsDocDefinition {
  anchorId: string;
  children?: SkyManifestChildDefinition[];
  docsId: string;
  docsIncludeIds?: string[];
  filePath: string;
  isInternal?: boolean;
  kind: SkyManifestParentDefinitionKind;
  name: string;
}

/**
 * Information shared by all child entities exported from the public API.
 * @internal
 */
export interface SkyManifestChildDefinition extends SkyManifestJsDocDefinition {
  kind: SkyManifestChildDefinitionKind;
  name: string;
  type: string;
}
