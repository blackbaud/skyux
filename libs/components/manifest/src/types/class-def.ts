import type {
  SkyManifestChildDefinition,
  SkyManifestParentDefinition,
} from './base-def';
import type { SkyManifestParameterDefinition } from './function-def';

/**
 * Information about a class exported from the public API.
 * @internal
 */
export interface SkyManifestClassDefinition
  extends SkyManifestParentDefinition {
  children?: (
    | SkyManifestClassMethodDefinition
    | SkyManifestClassPropertyDefinition
  )[];
  kind: 'class' | 'module' | 'service';
  typeParameters?: string;
}

/**
 * Information about a class method.
 * @internal
 */
export interface SkyManifestClassMethodDefinition
  extends SkyManifestChildDefinition {
  isStatic?: boolean;
  kind: 'class-method';
  parameters?: SkyManifestParameterDefinition[];
  typeParameters?: string;
}

/**
 * Information about a class property.
 * @internal
 */
export interface SkyManifestClassPropertyDefinition
  extends SkyManifestChildDefinition {
  defaultValue?: string;
  isStatic?: boolean;
  kind: 'class-property';
}
