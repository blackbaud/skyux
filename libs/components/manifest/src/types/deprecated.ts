export interface SkyManifestDeprecatedDirective {
  deprecationReason?: string;
  isDeprecated: boolean;
  properties?: {
    name: string;
    deprecationReason?: string;
  }[];
  selector: string;
}

export interface SkyManifestDeprecatedPipe {
  deprecationReason?: string;
  templateBindingName: string;
}

/**
 * Information about deprecated template features (directives, components, and pipes).
 */
export interface SkyManifestTemplateFeatureDeprecations {
  components: SkyManifestDeprecatedDirective[];
  directives: SkyManifestDeprecatedDirective[];
  pipes: SkyManifestDeprecatedPipe[];
}
