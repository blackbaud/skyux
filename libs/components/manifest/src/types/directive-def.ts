import type {
  SkyManifestChildDefinition,
  SkyManifestParentDefinition,
} from './base-def';

/**
 * Information about a directive or component exported from the public API.
 * @internal
 */
export interface SkyManifestDirectiveDefinition
  extends SkyManifestParentDefinition {
  children?: (
    | SkyManifestDirectiveInputDefinition
    | SkyManifestDirectiveOutputDefinition
  )[];
  kind: 'directive' | 'component';
  selector?: string;
}

/**
 * Information about a directive input property.
 * @internal
 */
export interface SkyManifestDirectiveInputDefinition
  extends SkyManifestChildDefinition {
  defaultValue?: string;
  isRequired?: boolean;
  kind: 'directive-input';
}

/**
 * Information about a directive input property.
 * @internal
 */
export interface SkyManifestDirectiveOutputDefinition
  extends SkyManifestChildDefinition {
  kind: 'directive-output';
}
