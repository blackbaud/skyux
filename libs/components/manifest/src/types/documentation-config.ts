export interface SkyManifestDocumentationConfig {
  packages: Record<
    string,
    {
      groups: Record<string, { docsIds: string[]; primaryDocsId: string }>;
    }
  >;
}
