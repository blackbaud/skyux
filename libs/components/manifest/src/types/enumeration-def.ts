import type {
  SkyManifestChildDefinition,
  SkyManifestParentDefinition,
} from './base-def';

/**
 * Information about an enumeration exported from the public API.
 * @internal
 */
export interface SkyManifestEnumerationDefinition
  extends SkyManifestParentDefinition {
  children: SkyManifestEnumerationMemberDefinition[];
}

/**
 * Information about an enumeration member.
 * @internal
 */
export interface SkyManifestEnumerationMemberDefinition
  extends SkyManifestChildDefinition {
  kind: 'enum-member';
}
