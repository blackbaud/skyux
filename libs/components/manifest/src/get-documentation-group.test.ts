import { SkyManifestParentDefinition } from './types/base-def';
import { SkyManifestDocumentationConfig } from './types/documentation-config';
import {
  SkyManifestCodeExamples,
  SkyManifestPublicApi,
} from './types/manifest';

describe('get-documentation-group', () => {
  function setup(options: {
    codeExamples: SkyManifestCodeExamples;
    documentationConfig: SkyManifestDocumentationConfig;
    publicApi: SkyManifestPublicApi;
  }): void {
    jest.mock('../code-examples.json', () => options.codeExamples);
    jest.mock(
      '../documentation-config.json',
      () => options.documentationConfig,
    );
    jest.mock('../public-api.json', () => options.publicApi);
  }

  afterEach(() => {
    jest.resetAllMocks();
    jest.resetModules();
  });

  it('should return information about a documentation group', async () => {
    setup({
      codeExamples: {
        examples: {
          FooCodeExample: {
            componentName: 'FooCodeExample',
            files: {
              'example.component.ts': 'TS_CONTENTS',
              'example.component.html': 'HTML_CONTENTS',
            },
            importPath: '@skyux/code-examples',
            primaryFile: 'example.component.ts',
            selector: 'lib-foo',
          },
          IndicatorCodeExample: {
            componentName: 'IndicatorCodeExample',
            files: {
              'example.component.ts': 'TS_CONTENTS',
              'example.component.html': 'HTML_CONTENTS',
            },
            importPath: '@skyux/code-examples',
            primaryFile: 'example.component.ts',
            selector: 'lib-indicator',
          },
        },
      },
      documentationConfig: {
        packages: {
          '@skyux/core': {
            groups: {
              foo: {
                docsIds: [
                  'FooComponent',
                  'FooHarness',
                  'IndicatorComponent',
                  'FooCodeExample',
                ],
                primaryDocsId: 'FooComponent',
              },
              bar: {
                docsIds: ['BarComponent'],
                primaryDocsId: 'BarComponent',
              },
            },
          },
          '@skyux/indicators': {
            groups: {
              indicators: {
                docsIds: ['IndicatorComponent', 'IndicatorCodeExample'],
                primaryDocsId: 'IndicatorComponent',
              },
            },
          },
        },
      },
      publicApi: {
        packages: {
          '@skyux/code-examples': [
            {
              name: 'FooCodeExample',
              docsId: 'FooCodeExample',
              extraTags: {
                title: 'My code example',
              },
              repoUrl: 'https://repo.com/foo',
            },
            {
              name: 'IndicatorCodeExample',
              docsId: 'IndicatorCodeExample',
              repoUrl: 'https://repo.com/foo',
            },
          ] as SkyManifestParentDefinition[],
          '@skyux/core': [
            {
              name: 'FooComponent',
              docsId: 'FooComponent',
              repoUrl: 'https://repo.com/foo',
            },
            {
              name: 'BarComponent',
              docsId: 'BarComponent',
              repoUrl: 'https://repo.com/foo',
            },
          ] as SkyManifestParentDefinition[],
          '@skyux/core/testing': [
            {
              name: 'FooHarness',
              docsId: 'FooHarness',
              repoUrl: 'https://repo.com/foo',
            },
          ] as SkyManifestParentDefinition[],
          '@skyux/indicators': [
            {
              name: 'IndicatorComponent',
              docsId: 'IndicatorComponent',
              repoUrl: 'https://repo.com/foo',
            },
            {
              name: 'IndicatorCodeExample',
              docsId: 'IndicatorCodeExample',
              repoUrl: 'https://repo.com/foo',
            },
          ] as SkyManifestParentDefinition[],
        },
      },
    });

    const { getDocumentationGroup } = await import('./get-documentation-group');

    const result = getDocumentationGroup('@skyux/core', 'foo');

    expect(result).toMatchSnapshot();
  });

  it('should throw if package name unrecognized', async () => {
    setup({
      codeExamples: {
        examples: {},
      },
      documentationConfig: {
        packages: {},
      },
      publicApi: {
        packages: {},
      },
    });

    const { getDocumentationGroup } = await import('./get-documentation-group');

    expect(() =>
      getDocumentationGroup('@skyux/invalid', 'foo'),
    ).toThrowErrorMatchingSnapshot();
  });

  it('should throw if group name unrecognized', async () => {
    setup({
      codeExamples: {
        examples: {},
      },
      documentationConfig: {
        packages: {
          '@skyux/core': {
            groups: {},
          },
        },
      },
      publicApi: {
        packages: {},
      },
    });

    const { getDocumentationGroup } = await import('./get-documentation-group');

    expect(() =>
      getDocumentationGroup('@skyux/core', 'invalid'),
    ).toThrowErrorMatchingSnapshot();
  });

  it('should throw if docsId unrecognized', async () => {
    setup({
      codeExamples: {
        examples: {},
      },
      documentationConfig: {
        packages: {
          '@skyux/core': {
            groups: {
              foo: {
                docsIds: ['invalid'],
                primaryDocsId: 'invalid',
              },
            },
          },
        },
      },
      publicApi: {
        packages: {},
      },
    });

    const { getDocumentationGroup } = await import('./get-documentation-group');

    expect(() =>
      getDocumentationGroup('@skyux/core', 'foo'),
    ).toThrowErrorMatchingSnapshot();
  });
});
