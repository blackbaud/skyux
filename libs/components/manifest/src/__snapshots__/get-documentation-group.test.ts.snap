// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`get-documentation-group should remove internal parents and children 1`] = `
Object {
  "codeExamples": Array [],
  "packageInfo": Object {
    "packageName": "@skyux/core",
    "registryUrl": "https://www.npmjs.com/package/@skyux/core",
    "repoUrl": "https://repo.com/foo",
  },
  "publicApi": Array [
    Object {
      "children": Array [
        Object {
          "name": "publicProperty",
        },
      ],
      "docsId": "FooComponent",
      "name": "FooComponent",
      "packageName": "@skyux/core",
      "repoUrl": "https://repo.com/foo",
    },
  ],
  "testing": Array [],
}
`;

exports[`get-documentation-group should return information about a documentation group 1`] = `
Object {
  "codeExamples": Array [
    Object {
      "componentName": "FooCodeExample",
      "files": Object {
        "example.component.html": "HTML_CONTENTS",
        "example.component.ts": "TS_CONTENTS",
      },
      "importPath": "@skyux/code-examples",
      "primaryFile": "example.component.ts",
      "selector": "lib-foo",
    },
  ],
  "packageInfo": Object {
    "packageName": "@skyux/core",
    "registryUrl": "https://www.npmjs.com/package/@skyux/core",
    "repoUrl": "https://repo.com/foo",
  },
  "publicApi": Array [
    Object {
      "docsId": "FooComponent",
      "name": "FooComponent",
      "packageName": "@skyux/core",
      "repoUrl": "https://repo.com/foo",
    },
    Object {
      "docsId": "IndicatorComponent",
      "name": "IndicatorComponent",
      "packageName": "@skyux/indicators",
      "repoUrl": "https://repo.com/foo",
    },
  ],
  "testing": Array [
    Object {
      "docsId": "FooHarness",
      "name": "FooHarness",
      "packageName": "@skyux/core/testing",
      "repoUrl": "https://repo.com/foo",
    },
  ],
}
`;

exports[`get-documentation-group should throw if docsId unrecognized 1`] = `"Failed to retrieve type definition with docsId \\"invalid\\"."`;

exports[`get-documentation-group should throw if group name unrecognized 1`] = `"Documentation exists for a package named \\"@skyux/core\\", but it does not include a group with the name \\"invalid\\"."`;

exports[`get-documentation-group should throw if package name unrecognized 1`] = `"Documentation could not be found for the package \\"@skyux/invalid\\"."`;
